server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/CachedMTreeReadWriteLock.java

```diff
@@ -0,0 +1,232 @@
+ *       i.e., if you only call lock() once then do not call unlock() more than once and vice versa.
+ *   <li>In particular, if the current thread has already requested a WriteLock, it can continue to
+ *       reenter to acquire a WriteLock or a ReadLock. However, if a ReadLock has already been
+ *       acquired (including by the current thread), the WriteLock acquiring will be blocked, which
+ *       may cause deadlocks. Therefore, if a write operation needs to be performed, a WriteLock
+ *       must be applied first.
+ *   <li>Writer preferred is used to avoid starving the write lock request, which means once a write
+ *       lock request is waiting, it will block the newly arriving read lock request expect for the
+ *       re-entrant read lock request.
+ * </ul>
+ */
+public class CachedMTreeReadWriteLock {
+  private final Lock lock = new ReentrantLock();
+  private final Condition okToRead = lock.newCondition();
+  private final Condition okToWrite = lock.newCondition();
+  private long stampAllocator = 0;
+
+  private Thread exclusiveOwnerThread;
+  private final Map<Long, Integer> readCnt = new HashMap<>();
+  private int readWait = 0;
+  private int writeCnt = 0;
+  private int writeWait = 0;
+
+  private final ThreadLocal<Long> sharedOwnerStamp = new ThreadLocal<>();
+
+  public static final long NON_STAMP = -1;
+  public static final long ALLOCATE_STAMP = -2;
+
+  /**
+   * Acquires the read lock.
+   *
+   * @param stamp NON_STAMP, ALLOCATE_STAMP or re-entry based on stamp
+   *     <p>NON_STAMP means read lock acquire and release is thread-bound and supports re-entry
+   *     within the same thread. Return directly if no thread holds a write lock or the current
+   *     thread already holds a write lock; block and wait if another thread holds a write lock or
+   *     the write lock waiting queue is not empty.
+   *     <p>ALLOCATE_STAMP means read lock acquire and release is stamp-bound and supports re-entry
+   *     by the same stamp. Return a new stamp if no thread holds a write lock; block and wait if
+   *     another thread holds a write lock or the write lock waiting queue is not empty.
+   *     <p>Re-entry based on stamp. If the stamp exists, increase the read count and return;
+   *     otherwise, same as ALLOCATE_STAMP.
+   * @return stamp or NON_STAMP
+   */
+  public long readLock(long stamp) {
+    lock.lock();
+    try {
+      if (stamp == NON_STAMP) {
+        return threadReadLock();
+      } else if (stamp == ALLOCATE_STAMP) {
+        return acquireReadLock();
+      } else {
+        if (readCnt.containsKey(stamp)) {
+          readCnt.put(stamp, readCnt.get(stamp) + 1);
+          return stamp;
+        } else {
+          return acquireReadLock();
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private long threadReadLock() {
+    Long allocateStamp = sharedOwnerStamp.get();
+    if (allocateStamp == null) {
+      // first time entry, acquire read lock and set thread local
```

Comment:
We can set special interfaces for first time entry, rather than mix it with reentrant cases.

Reply:
done